lib/ecto/repo/model.ex

```diff
@@ -182,11 +181,32 @@ defmodule Ecto.Repo.Model do
     put_in model.__meta__.state, :loaded
   end
 
-  defp merge_into_changeset(struct, fields, changeset) do
-    changes = Map.take(struct, fields)
+  defp merge_into_changeset(struct, fields, embeds, changeset) do
+    changes =
+      struct
+      |> Map.take(fields)
+      |> Map.drop(embeds)
+
     update_in changeset.changes, &Map.merge(changes, &1)
   end
 
+  defp apply_embedded_callbacks(embeds, changeset, callback) do
+    types = changeset.types
+
+    update_in changeset.changes, fn changes ->
+      Enum.reduce(embeds, changes, fn name, changes ->
+        case Map.fetch(changes, name) do
+          {:ok, changeset} ->
+            {:embed, embed} = Map.get(types, name)
+            Map.put(changes, name, Ecto.Embedded.apply_callback(embed, changeset, callback))
```

Comment:
I think it will be better to just pass if its before or after callback, and take action from the changeset.